### DISCUSSION TITLE
Allow changes in behaviour while retaining backward compatibility

### DISCUSSION THREAD

**Discussion by mberndt123** [sbt/sbt#7582] [Ideas]
Hey there,

As a program like sbt evolves over time, we often find better ways to do things. But these new ways are often incompatible with how things have worked thus far, and we can't make the new way the default because that would break everybody's projects.
As an example, when I tried to publish an sbt plugin to my Nexus, I got an error and I had to set the somewhat obscure `sbtPluginPublishLegacyMavenStyle := false` flag to make that work.

I therefore propose to add a compatibility version setting that tells sbt the version that we expect its behaviour to be compatible with. For example, when an incompatible behaviour change to a task is introduced in sbt 1.11.0 and the compatibility version is set to 1.10.0, then that task will behave in the 1.10.0 way. Set the compatibility version to 1.11.0 and you get the new behaviour. The advantage over just using the old version is that you still get to benefit from bug fixes, optimizations etc., while not having to update your build at an inconvenient time.

There is ample precedent for such features. In CMake, the [`cmake_policy`](https://cmake.org/cmake/help/latest/command/cmake_policy.html) or [`cmake_minimum_required`](https://cmake.org/cmake/help/v3.5/command/cmake_minimum_required.html) command is used for this purpose. Perl has the [`use VERSION`](https://perldoc.perl.org/functions/use#use-VERSION) feature. NixOS has the [`system.stateVersion`](https://search.nixos.org/options?channel=24.05&show=system.stateVersion) setting.

The compatibility version setting should by default be the same as the sbt version, but a warning should be printed on startup if it is not set explicitly, much like CMake prints a warning if `cmake_minimum_required` is not used. The granularity of this setting is to be discussed: is it just a global thing, or should it be possible to override it per project or task?

Sbt Plugins should define their own compatibility version setting if they deem it necessary.

**Comment by eed3si9n** [comment]
> I therefore propose to add a compatibility version setting that tells sbt the version that we expect its behaviour to be compatible with. For example, when an incompatible behaviour change to a task is introduced in sbt 1.11.0 and the compatibility version is set to 1.10.0, then that task will behave in the 1.10.0 way.

When you say "I therefore propose", are you proposing to work on this for all future versions of sbt? I don't know if promising some behavioral compatibility is realistic when underlying projects like Couriser or JLine could also introduce breaking changes.

**Reply by mberndt123** [reply]
Well you're never going to achieve 100% compatibility. After all, every possible change is going to break somebody's workflow in some way.
https://xkcd.com/1172/
But I'd argue that the example of CMake demonstrates that this approach can take us pretty far.

**Reply by som-snytt** [reply]
"mberndt123 proposes, eed3si9n disposes".

I never use obscure sbt settings, but `-source` for scalac serves a similar purpose: usually, enable "future" features, but Scala 3 uses it for backward compatibility to assist migration. It is a relaxed guarantee for a dissolute workflow.

**Reply by eed3si9n** [reply]
Given that sbt 1.x keeps bincompat, the scalac equivalent is we are talking about `-Xsource2.13.12` here. When sbt 2.x comes out, and if some plugins cross build like they did for 0.13/1.x then I can see some cross emulation, but again I'm not sure who would work on that.

**Comment by rmgk** [comment]
This seems to imply that there are a lot of settings where nearly everyone would want to change the default. But I question if this is the case. None of my sbt projects changes any defaults in a non-project specific way. People with similar projects with shared configuration seem to use sbt plugins.

The only example you mention is linked to Bintray shutting down if I remember correctly, which is quite a rare event.

**Reply by mberndt123** [reply]
There are more examples. For example, every change of the Scala version is an incompatible change, even patch versions, because Scala compiler plugins need to match even the patch version of the scala compiler. So if you add a scala compiler plugin in your build and then increase the scala compiler version, your build will probably break.
Also what about "turbo" mode, added in sbt 1.3.0 but disabled by default? What about `usePipelining`? These should arguably be the default. I also think that Scala 3 should be the default at this point.
